### DISCUSSION TITLE
[5.x]: Rendering custom fields doesn't work

### DISCUSSION THREAD

**Discussion by mikesnoeren** [solspace/craft-freeform#1074] [Questions & Answers] [answered]
In 4.x, we were able to render custom fields using:

```twig
{{ field.render({
    labelClass: 'block ' ~ (field.required ? 'required'),
    class: "mt-1 mb-2 w-full rounded border-neutral-400",
}) }}
```

Even though this is still in the docs (outdated?), this currently doesn't work as it ignores all given properties and values..
Instead, we have to do this:

```twig
<div class="flex flex-col" data-field-container="{{ field.handle }}" data-field-type="{{ field.type }}">
    <label for="form-input-{{ field.handle }}">{{ field.label|t }}</label>
    <input name="{{ field.handle }}" type="{{ field.type }}" id="form-input-{{ field.handle }}" value="{{ field.value ?? '' }}" class="mt-1 mb-2 w-full rounded border-neutral-400">
</div>
```

Which is a lot less pretty.

**Comment by kjmartens** [comment, marked as answer]
Hi @mikesnoeren,

Sorry about the old code suggestion in the documentation. We are working to clean up the remaining things and that should be done shortly.

In the meantime, you can refer to the docs here:
https://docs.solspace.com/craft/freeform/v5/templates/formatting/

This can be achieved at the form level (see `fields` -> `@global` AND `fields` -> `:required`):

``` twig
{{ form.render({
    attributes: {
        novalidate: true,
        row: {
            class: "row",
        },
    },
    fields: {
        "@global": {
            attributes: {
                input: {
                    class: "input-class",
                },
                container: {
                    "data-container": true,
                },
                label: {
                    class: "label-class",
                },
            },
        },
        "@dropdown": {
            attributes: {
                input: {
                    class: "select fullwidth",
                },
                container: {
                    "data-select-container": true,
                },
            },
        },
        "@checkboxes": {
            instructions: "Select all that apply.",
        },
        "@text": { ... },
        "@checkboxes, @multiple-select": {
            instructions: "Select all that apply.",
        },
        "myFieldHandle": {
            attributes: {
                container: {
                    "data-my-input-container": true,
                },
                input: {
                    id: "my-field",
                    class: "big-text",
                },
            },
        },
        ":options": { ... },
        ":required": { ... },
        ":errors": { ... },
        ":required, @dropdown, @text, myFieldHandle": { ... }
    },
}) }}
```

Or it can be done at the [field level](https://docs.solspace.com/craft/freeform/v5/templates/formatting/#array-values-conditionals):

``` twig
{{ field.render({
    attributes: {
        label: {
            id: "label-for-" ~ field.handle,
            class: [
                "form-control-label",
                field.hasErrors ? "is-invalid has-validation",
                field.required ? "required",
                field.type == "checkbox" ? "checkbox" : "non-checkbox",
            ],
        },
    },
}) }}
```

**Comment by MalcolmJohnston** [comment]
The above has really helped me with migrating some formatting templates that are heavy on separately rendered fields.

One issue I have not got around yet is how to set the error class using field.renderErrors?

I thought it would be as follows:

```
{{ field.renderErrors({
    attributes: {
        "@global": {
            class: [
                "my_error_class",
            ],
        },
    },
}) }}
```

But that doesn't seem to do it?

Cheers,
Malcolm

**Reply by gustavs-gutmanis** [reply]
Here you would do this instead:

```
{{ field.renderErrors({
    attributes: {
        error: {
            class: "my_error_class",
        },
    },
}) }}
```

But you can do this earlier, in the `field.render` as well, or during `form.render` if you want this to happen for all field error messages.

If you do this in `form.render` though, you would use the `@global` attribute

```
form.render({
  fields: {
    "@global": {
      attributes: {
        error: {
          class: "my_error_class",
        },
      },
    },
  },
})
```

**Reply by mikesnoeren** [reply]
It seems like the error classes don't get applied when the form is submitted using AJAX, can this be the case? I tried your exact example here, and it only works when AJAX is off. Is this intended behaviour?

**Reply by gustavs-gutmanis** [reply]
Yes, that is unfortunately the case.
It happens because AJAX forms receive errors as an array of strings from the response JSON and then generate the error elements.

It uses the class names assigned via the JS options passed to the form JS plugin.

You can override them with this JS snippet:

```html
<script>
  const form = document.getElementById('my-form');
  form.addEventListener('freeform-ready', function(event) {
    event.options.errorClassField = 'my_error_class';
  });
</script>
```

You can read about more options [here](https://docs.solspace.com/craft/freeform/v5/developer/js-plugin/#plugin-options).

We might be able to make it so that the default options for these error classes and success banners are extracted from the Form Render Options. We will add a ticket to look into this.

**Reply by mikesnoeren** [reply]
Thanks for getting back to me so quickly.

Unfortunately, when using tailwindCSS, you often have to apply conditional styles like this: `!bg-red-500`.  
When we apply that to your given example we get an JS error.
```
form.addEventListener('freeform-ready', function(event) {
    event.options.errorClassField = '!bg-red-500';
});
```
`Uncaught DOMException: Failed to execute 'querySelectorAll' on 'Element': '.!hidden' is not a valid selector`

Removing the exclamation mark fixes the problem, but is likely not a good option for tailwindCSS users.

**Reply by gustavs-gutmanis** [reply]
That is very unfortunate. We are also using the same class to look up the element to remove it after there are no errors.

There is the option to override the way errors are rendered [described here](https://docs.solspace.com/craft/freeform/v5/developer/js-plugin/#rendering-field-errors-freeform-render-field-errors) which would let you reign free with your custom implementations.

Also there's an [event for removing the error messages](https://docs.solspace.com/craft/freeform/v5/developer/js-plugin/#removing-field-messages-freeform-remove-field-messages), which you would have to use to remove your custom made error message displaying.

Hope this helps.

We will try to update the logic to be less edge case error prone, and more generic, to simplify it for our users.

**Reply by mikesnoeren** [reply]
Thanks for getting back to me so quickly.

I ended up settling for this CSS approach that works fine.
Ideally, the AJAX script applies the given error classes for us. 
But this will do for now.

```css
.freeform-has-errors {
    @apply !border-red-500;
}

.freeform-errors {
    @apply text-red-500;
}
```

**Reply by gustavs-gutmanis** [reply]
Glad it worked out!

We'll think of ways to address this problem in the future.
